pep.rst

```diff
@@ -0,0 +1,295 @@
+alleviates the typical portability issue by simply not making it the
+responsibility of CPython to provide a JIT that works across supported
+platforms. Exposing an API instead of using a specific JIT with
+CPython also allows for the best JIT to be used per platform and
+workload. Finally, it allows for easier experimentation with CPython
+by JIT authors. This could help facilitate an ecosystem of JITs for
+CPython.
+
+The second part of our solution for overcoming traditional objections
+at adding a JIT to CPython is to expose a *small* expansion of the C
+API of CPython with a *simple* entrypoint at which CPython's eval loop
+interacts with any potential JIT. By keeping the semantics of the
+proposed JIT API small and simple it is easy to comprehend for both
+CPython contributors and JIT authors as well as making the maintenance
+burden extremely small on CPython contributors themselves.
+
+
+Proposal
+========
+
+Expanding ``PyCodeObject``
+--------------------------
+
+Two new fields are to be added to the ``PyCodeObject`` struct
+[#pycodeobject]_ along with a constant::
+
+  typedef struct {
+     ...
+     PyJittedCode *co_jitted;  /* Stores struct to JIT details. */
+     PY_UINT64_T co_run_count;  /* The number of times the code object has run. */
+  } PyCodeObject;
+
+  /* Constant to represent when a JIT cannot compile a code object;
+     should not compare equal to NULL. */
+  const PyJittedCode *PY_JIT_FAILED = 1;
+
+The ``co_jitted`` field stores a pointer to a ``PyJittedCode`` struct
+which stores the details of the compiled JIT code (``PyJittedCode`` is
+explained later). By adding a level of indirection for JIT compilation
+details per code object, the memory impact of this API is minimized
+when a JIT is not used/triggered.
+
+The ``co_run_count`` field counts the numbers of executions for the
+code object. This allows for JIT compilation to only be triggered for
+"hot" code objects instead of all code objects. This helps mitigate
+JIT compilation overhead by only triggering compilation for heavily
+used code objects. (If issue #26219 [#26219]_ is accepted then the run
+count per code object will already be present in CPython).
+
+The ``PY_JIT_FAILED`` constant is used to signify when a JIT attempted
+to compile a code object and failed. Setting this constant prevents
+attempting to compile the code object in the future. The value should
+be a memory address that has nearly no chance of ever being a valid
+memory address to a ``PyJittedCode`` struct.
+
+
+``PyJittedCode``
+----------------
+
+The ``PyJittedCode`` struct carries all relevant details about what
+the JIT produced for the code object it is attached to::
+
+  typedef PyObject* (__stdcall*  Py_EvalFunc)(void*, struct _frame*);
+
+  typedef struct {
+      PyObject_HEAD
+      Py_EvalFunc jit_evalfunc;
+      void *jit_evalstate;
+  }
+
+The ``jit_evalfunc`` points to a trampoline function provided by the
```

Comment:
This isn't strictly a trampoline function...  I would say:

The ``jit_evalfunc`` points to a function provided by the JIT to handle execution of a Python frame and returning the resulting ``PyObject`` (much like ``PyEval_EvalFrameEx()``  [#pyeval_evalframeex]_).  The function can point directly to the generated code or can be a trampoline function.  The JIT may use trampoline functions to provide tracing and call back into the Python byte code interpreter.  Trampolines can also be used to perform common prologue and epilogue tasks rather than emitting it into generated code.